pkg/kademlia/kademlia.go

```diff
@@ -5,243 +5,198 @@ package kademlia
-		}
-	}()
-
-	return nil
-}
+	// begin the work looking for the node by spinning up alpha workers
+	// and asking for the node we want from nodes we know in our routing table
+	ch := make(chan []*proto.Node, conc)
+	w := newWorker(ctx, k.routingTable, nodes, k.nodeClient, target, kb)
+	ctx, w.cancel = context.WithCancel(ctx)
+	defer w.cancel()
+	for i := 0; i < conc; i++ {
+		go w.work(ctx, ch)
+	}
 
-func convertProtoNodes(n []proto.Node) ([]*bkad.NetworkNode, error) {
-	nn := make([]*bkad.NetworkNode, len(n))
-	for i, v := range n {
-		node, err := convertProtoNode(v)
-		if err != nil {
-			return nil, err
+	for {
+		select {
+		case v := <-ch:
+			for _, node := range v {
+				if node.GetId() == target.String() {
+					return node, nil
+				}
+			}
+		case <-ctx.Done():
+			return nil, NodeNotFound
 		}
-		nn[i] = node
 	}
-
-	return nn, nil
 }
 
-func convertNetworkNodes(n []*bkad.NetworkNode) []*proto.Node {
-	nn := make([]*proto.Node, len(n))
-	for i, v := range n {
-		nn[i] = convertNetworkNode(v)
-	}
-
-	return nn
+// Ping checks that the provided node is still accessible on the network
+func (k *Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
+	// TODO(coyle)
+	return proto.Node{}, errors.New("TODO")
 }
 
-func convertNetworkNode(v *bkad.NetworkNode) *proto.Node {
-	return &proto.Node{
-		Id:      string(v.ID),
-		Address: &proto.NodeAddress{Transport: defaultTransport, Address: net.JoinHostPort(v.IP.String(), strconv.Itoa(v.Port))},
-	}
+// FindNode looks up the provided NodeID first in the local Node, and if it is not found
+// begins searching the network for the NodeID. Returns and error if node was not found
+func (k *Kademlia) FindNode(ctx context.Context, ID dht.NodeID) (proto.Node, error) {
+	//TODO(coyle)
+	return proto.Node{}, NodeErr.New("TODO")
 }
 
-func convertProtoNode(v proto.Node) (*bkad.NetworkNode, error) {
-	host, port, err := net.SplitHostPort(v.GetAddress().GetAddress())
-	if err != nil {
-		return nil, err
-	}
-
-	nn := bkad.NewNetworkNode(host, port)
-	nn.ID = []byte(v.GetId())
-
-	return nn, nil
-}
+// ListenAndServe connects the kademlia node to the network and listens for incoming requests
+func (k *Kademlia) ListenAndServe() error {
+	//TODO(coyle)
 
-// newID generates a new random ID.
-// This purely to get things working. We shouldn't use this as the ID in the actual network
-func newID() ([]byte, error) {
-	result := make([]byte, 20)
-	_, err := rand.Read(result)
-	return result, err
+	return errors.New("TODO")
 }
 
 // GetIntroNode determines the best node to bootstrap a new node onto the network
-func GetIntroNode(id, ip, port string) (*proto.Node, error) {
-	addr := "bootstrap.storj.io:8080"
-	if ip != "" && port != "" {
-		addr = ip + ":" + port
-	}
-
-	if id == "" {
-		i, err := newID()
-		if err != nil {
-			return nil, err
-		}
-
-		id = string(i)
+func GetIntroNode(addr string) (*proto.Node, error) {
+	if addr == "" {
+		addr = "bootstrap.storj.io:8080"
```

Comment:
we should make this a flag

Reply:
also, we should not put production values in to the codebase imo. otherwise we run the risk of having unit tests join the production network. at one company, we 3 separate times ended up charging customers' credit cards due to integration testing.

Reply:
Do we still need this? I'm thinking we just pull intro node directly from config. Thoughts?

Reply:
sgtm

Reply:
https://storjlabs.atlassian.net/browse/V3-434